File: 백준/Gold/2470. 두 용액/두 용액.java
import java.io.*;
import java.util.*;

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;

		int N = Integer.parseInt(br.readLine());
		int[] arr = new int[N];

		st = new StringTokenizer(br.readLine(), " ");
		for (int i = 0; i < N; i++)
			arr[i] = Integer.parseInt(st.nextToken());

		Arrays.sort(arr);

		int diff = Integer.MAX_VALUE;
		int s = 0;
		int e = N - 1;

		int ans1 = 0;
		int ans2 = 0;

		while (s < e) {
			int temp = arr[s] + arr[e];
			int absValue = Math.abs(temp);

			if (absValue < diff) {
				diff = absValue;
				ans1 = arr[s];
				ans2 = arr[e];
				if (absValue == 0)
					break;
			}
			if (temp > 0)
				e-=1;
			else
				s+=1;
		}
		System.out.println(ans1 + " " + ans2);
	}

}

File: 백준/Gold/2470. 두 용액/두 용액.py
N = int(input())
arr = list(map(int, input().split()))
arr.sort()

s = 0
e = len(arr)-1
diff = abs(arr[s] + arr[e])
ans = [arr[s], arr[e]]

while s < e:
    left = arr[s]
    right = arr[e]

    if abs(left+right) < diff:
        diff = abs(left+right)
        ans = [left, right]
        if diff == 0:
            break
    if (left+right) < 0:
        s += 1
    else:
        e -= 1


print(ans[0], ans[1])
